### ISSUE TITLE
Spring Boot health check fails with a user-provided RestHighLevelClient as the health indicator continues to use the auto-configured RestClient

### ISSUE THREAD

**Issue by sudoapt-getclean** [spring-projects/spring-boot#17464]
In my `@SpringBootApplication` class I define this restclient

```java
    @Bean(destroyMethod = "close")
    public RestHighLevelClient client() {
        ClientConfiguration clientConfiguration = ClientConfiguration.builder()
            .connectedTo(elasticsearchHost)
            .usingSsl()
            .build();

        return RestClients.create(clientConfiguration).rest();
    }
```
This all works fine. however when the /actuator/health endpoint of the application is called. It fails and I get a stacktrace from elastic that shows connection refused.
Some debugging showed me that this is because I have two RestClient beans. The second bean is created by the code in RestClientAutoconfiguration below. This is done before my RestClient is created. 
This other Restclient is created with the default localhost:9200 configuration and is used for the health check (but not for the actual searching)

```java
/**
 * {@link EnableAutoConfiguration Auto-configuration} for Elasticsearch REST clients.
 *
 * @author Brian Clozel
 * @since 2.1.0
 */
@Configuration(proxyBeanMethods = false)
@ConditionalOnClass(RestClient.class)
@EnableConfigurationProperties(RestClientProperties.class)
public class RestClientAutoConfiguration {

	@Bean
	@ConditionalOnMissingBean
	public RestClient restClient(RestClientBuilder builder) {
		return builder.build();
	}
```
I happen to be running spring boot 2.2.0.M4 (Because of the support for using the highlevel restclient yay)
Keep up the good work and let me know If you need more information

--- Appendix ---
the original stacktrace is
```
2019-07-09 13:54:19.981 DEBUG 1 --- [io-8080-exec-10] o.s.web.servlet.DispatcherServlet        : GET "/actuator/health", parameters={}
2019-07-09 13:54:20.000  WARN 1 --- [io-8080-exec-10] s.b.a.e.ElasticsearchRestHealthIndicator : Elasticsearch health check failed

java.net.ConnectException: Connection refused
	at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:959) ~[elasticsearch-rest-client-6.7.2.jar!/:6.7.2]
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:233) ~[elasticsearch-rest-client-6.7.2.jar!/:6.7.2]
```

**Comment by sudoapt-getclean** [comment]
I since just discovered that If I set:
`spring.elasticsearch.rest.uris=` to the same host name as used in my configuration the problem goes away (as is to be expected because both beans now point to the correct host.)
however the problem is still not solved since the host I use is ssl and `org.springframework.boot.autoconfigure.elasticsearch.rest.RestClientProperties` does not offer support for that. 
so the message than is `<center>The plain HTTP request was sent to HTTPS port</center>`
for my particular usecase it would be great if if I could just configure my client with
`
spring.elasticsearch.rest.uris=hostname:443
spring.elasticsearch.rest.ssl=true
`
If other people however use more variables when building their `RestHighLevelClient` it is still necessary to hold off on autoconfiguration until after there has been a chance to create those. (Although there isn't that much to configure except for socket an connection timeouts and header.)

**Comment by wilkinsona** [comment]
https://github.com/spring-projects/spring-boot/issues/16049 is somewhat similar to this in that there was some problems there with the difference between the `RestClient` and the `RestHighLevelClient`.

**Comment by sudoapt-getclean** [comment]
Hey @wilkinsona thnx for the pompt reply!
I just added an extra bean definition
```java

    @Bean(destroyMethod = "close")
    public RestClient Healthclient() {
        ClientConfiguration clientConfiguration = ClientConfiguration.builder()
            .connectedTo(elasticsearchHost)
            .usingSsl()
            .build();

        return RestClients.create(clientConfiguration).lowLevelRest();
    }
```
and this indeed fixes the problem.
I now see that I can also reuse the (LowLevel)Restclient from HighLevelRestClient
```java
    @Bean
    public RestClient Healthclient() {
        return client().getLowLevelClient();
    }
```
but I'm not sure if this is wise, It actually looks good to me. so I'm just gonna try it.

It's up to you guys to decide if this requires code changes. the fact that the above fixes this kind of indicates to me that that could be done nicely inside the library somewhere.

Anyway Thanx again!

**Comment by wilkinsona** [comment]
Thanks for letting us know that you've solved your problem. I agree that there may be some room for improvement here. Let's re-open this one so that we don't forget to see what we can do.

**Comment by sudoapt-getclean** [comment]
I just read you title change again and I'm wondering if it would be correct for me to say that:

Perhaps there shouldn't be an auto-configured rest-client at all if the user creates one inside the HighLevelRestClient

**Comment by wilkinsona** [comment]
Yeah, that may make sense. The health indicator could then use the low-level client from the high-level client.

**Comment by bclozel** [comment]
The team discussed this issue. We came to the conclusion that the current auto-configuration is flawed, since developers can provide their own `RestHighLevelClient` bean (entirely configured by hand) and that the auto-configuration will contribute still a `RestClient` that's relying on the defaults.

We can fix that situation by making sure that:

* the `RestClient` bean should be created from a `RestHighLevelClient` if such a bean has been contributed by the application, or it should be auto-configured using the configuration properties (like the current situation)
* the `RestHighLevelClient` auto-configuration should remain the same: Spring Boot should contribute one only if the application is not already providing one.
* the health indicator doesn't have to change, since injecting the `RestClient` should get the right information

**Comment by bclozel** [comment]
Closed in favour of PR #17488
